lib/system.nim

```diff
@@ -2542,8 +2542,40 @@ template spliceImpl(s, a, L, b: untyped): untyped =
   # fill the hole:
   for i in 0 ..< b.len: s[a+i] = b[i]
 
-template `^^`(s, i: untyped): untyped =
-  (when i is BackwardsIndex: s.len - int(i) else: int(i))
+template `^^`*(s: typed; i: BackwardsIndex): int =
+  ## General converter template to convert the Backwards index back to
+  ## a normal integer index. If your own type has a special conversion
+  ## from Backwards index to int, you can write your own overload,
+  ## otherwis it will be just default to this: `s.len - int(i)`.
+  ##
+  ## .. code-block:: nim
+  ##
+  ##   type
+  ##     MyType = object
+  ##
+  ##   template `^^`(a: MyType, b: BackwardsIndex): int =
+  ##     ## my conversion of `BackwardsIndex` for `MyType`
+  ##     1000 + int(b)
+  ##
+  ##   proc `[]`(arg: MyType; idx: int): int =
+  ##     idx
+  ##
+  ##   var mt: MyType
+  ##   echo mt[^1] # 1001
+  s.len - int(i)
+
+template `^^`*[Idx, T](a: array[Idx, T], i: BackwardsIndex): int =
+  ## Converter template specialized for the array type.
+  int(a.low) + a.len - int(i)
+
+template `^^`*[Idx, T](a: array[Idx, T], i: Idx): int =
+  ## Converter template specialized for the array type.
+  int(i)
```

Comment:
Shouldn't these emit `Idx`, or else it won't work with non-int based array.

Reply:
it needs to be `int`, otherwise you can't subtract it from `len`. And yes backwards indexing doesn't work for `non-int` based array. But why should it, forward indexing `[0]` also doesn't work.

Reply:
But you can convert it back after you're done subtracting? It's how the original implementation for arrays work, too.

```nim
type
  E = enum
    A
    B
    C
    D

var x: array[E, int]

x[C] = 42
echo x[^2]
```

This compiles right now and won't with your version.

Reply:
The index at least would have to be an ordinal, then if it covers to a contiguous integer range we can subtract.

Otherwise, I suppose we could go out on a limb and say ordinal is a stable order for the collection itself, this isn't strictly true and hence my overall reservations. I do like what this is trying to do, but our type system needs to catch-up to the ambitions of the PR.

Reply:
A type system that would enable this PR without the issues would need a way to annotate collections to mark them as eglible for backwards indexing.
But we already have a way to mark collections as such, simply by implementing `\`[]\`(i: BackwardsIndex)`.
I don't think this PR is a good idea, the concept of backwards indexing is not universial and thus needs to be opt-in for collections and implementing the appropriate overloads is the simplest and most flexible (allows skipping len call in a linked list for example) way to opt-in.